packages/vitest/src/runtime/entry.ts

```diff
@@ -45,22 +52,31 @@ export async function run(files: string[], config: ResolvedConfig): Promise<void
     if (!files || !files.length)
       continue
 
-    await withEnv(environment, config.environmentOptions || {}, async () => {
-      for (const file of files) {
+    const filesByOptions = groupBy(files, ({ envOptions }) => JSON.stringify(envOptions))
```

Comment:
What about this case?

```ts
// file1.ts
// @vitest-environment custom
// @vitest-environment-options {"option1":1,"option2":2}

// file2.ts
// @vitest-environment custom
// @vitest-environment-options {"option2":2,"option1":1}
```

These files should run in the same environment, because they have the same values.

Reply:
Personally I don't think that's beneficial enough to be worth the downsides. We would have to make the equality logic much more complex. Also from what I understand, the JSON and ECMAScript spec do not specify whether parsers need to be ordered, so we would need to herd code different logic for every JavaScript runtime and environment.

If you wanted to ensure better performance, you could simple sort your kegs with a linter.

Reply:
> Also from what I understand, the JSON and ECMAScript spec do not specify whether parsers need to be ordered, so we would need to hard code different logic for JavaScript runtimes and environments.

I think you overcomplicate things here, I'm sure we can create a string key that doesn't depend on key order. Sorting keys seems like an obvious solution. 

> If you wanted to ensure better performance as a user, you could simply sort your keys with a linter. 

I don't think you can store them with linter in our docblock?

> Would something break if we made the wrong assumption about option equality for grouping envs?

Most things that expect single JSOM instance will break, because we don't clear ESM cache in `--no-threads`: Vue, for example, stores "document" globally on import, so next tests with different "options" will fail, because global "document" will be different. This is actually why we have grouping.

I am sure this usecase will emerge, if people will use `@vitet-environment-options`.

Reply:
> > Also from what I understand, the JSON and ECMAScript spec do not specify whether parsers need to be ordered, so we would need to hard code different logic for JavaScript runtimes and environments.
> 
> I think you overcomplicate things here, I'm sure we can create a string key that doesn't depend on key order. Sorting keys seems like an obvious solution.

That sounds like a good place to start, but I still feel like that's a lot of time complexity.

> > Would something break if we made the wrong assumption about option equality for grouping envs?
> 
> Most things that expect single JSOM instance will break, because we don't clear ESM cache in `--no-threads`: Vue, for example, stores "document" globally on import, so next tests with different "options" will fail, because global "document" will be different. This is actually why we have grouping.

So does that mean you can't have multiple tests with different options for the same env when using Vue with `--no-threads`?